registry/handlers/referrers.go

```diff
@@ -0,0 +1,336 @@
+package handlers
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"path"
+	"strconv"
+	"strings"
+
+	"github.com/distribution/distribution/v3"
+	dcontext "github.com/distribution/distribution/v3/context"
+	"github.com/distribution/distribution/v3/manifest/ociartifact"
+	"github.com/distribution/distribution/v3/registry/api/errcode"
+	v2 "github.com/distribution/distribution/v3/registry/api/v2"
+	"github.com/distribution/distribution/v3/registry/storage"
+	"github.com/distribution/distribution/v3/registry/storage/driver"
+	"github.com/gorilla/handlers"
+	"github.com/opencontainers/go-digest"
+	v1 "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+const (
+	// minimum page size used for # of digests to put in nextToken
+	minPageSize = 3
+	maxPageSize = 100
+)
+
+// referrersDispatcher takes the request context and builds the
+// appropriate handler for handling referrers requests.
+func referrersDispatcher(ctx *Context, r *http.Request) http.Handler {
+	dgst, err := getDigest(ctx)
+	if err != nil {
+
+		if err == errDigestNotAvailable {
+			return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				ctx.Errors = append(ctx.Errors, v2.ErrorCodeDigestInvalid.WithDetail(err))
+			})
+		}
+
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			ctx.Errors = append(ctx.Errors, v2.ErrorCodeDigestInvalid.WithDetail(err))
+		})
+	}
+
+	referrersHandler := &referrersHandler{
+		Context: ctx,
+		Digest:  dgst,
+	}
+	return handlers.MethodHandler{
+		"GET": http.HandlerFunc(referrersHandler.GetReferrers),
+	}
+}
+
+// referrersHandler handles http operations on referrers.
+type referrersHandler struct {
+	*Context
+	Digest digest.Digest
+}
+
+// ReferrersAPIResponse describes the response body of the referrers API.
+type ReferrersAPIResponse struct {
+	SchemaVersion int             `json:"schemaVersion"`
+	MediaType     string          `json:"mediaType"`
+	Manifests     []v1.Descriptor `json:"manifests"`
+}
+
+// GetReferrers fetches the list of referrers as an image index from the storage.
+func (h *referrersHandler) GetReferrers(w http.ResponseWriter, r *http.Request) {
+	dcontext.GetLogger(h).Debug("GetReferrers")
+
+	query := r.URL.Query()
+	n, nExists := query["n"]
+	var nPage int
+	var nParseError error
+	if !nExists {
+		nPage = maxPageSize
+	} else {
+		nPage, nParseError = strconv.Atoi(n[0])
+		// client specified nPage must be greater than min page size and less than or equal to max page size
+		if nParseError != nil || nPage < minPageSize || nPage > maxPageSize {
+			nPage = maxPageSize
+		}
+	}
+
+	nextTokens := query["nextToken"]
+	TokenSet := make(map[string]string)
```

Comment:
Since it is a set, value is never used.
```suggestion
	tokenSet := make(map[string]struct{})
```

Reply:
removed this part as it is part of the paging feature.